### ISSUE TITLE
Pulumi doesn't seem compatible with AGIC

### ISSUE THREAD

**Issue by lbargery** [pulumi/pulumi-azure#655]
Hey,

I'm trying to spin up a Kubernetes cluster (AKS) on Azure with an Application Gateway and then use something called [Application Gateway Ingress Controller](https://docs.microsoft.com/en-us/azure/application-gateway/ingress-controller-overview) to configure the app gateway in future. 

To do this, I've created the AKS cluster and a "blank" app gateway on Pulumi. Then was planning to hand off control to AGIC and ignore any changes to the Http listeners, backend pools etc. However, when I do a pulumi refresh and it brings in the new application gateway state, it seems like the business logic rules in Pulumi are unhappy with some the things that AGIC has done (even if I ignore) and so I get stuck when I do a pulumi preview (post the refresh). I get lots of the following errors
![image](https://user-images.githubusercontent.com/48315337/89550024-03e30d00-d801-11ea-8a12-e99d08e42b33.png)

For example, it seems like Pulumi requires a backend pool to have an IP address or FQDN set, but there is not a restriction Azure seems to impose (see attached).
![image](https://user-images.githubusercontent.com/48315337/89549162-db0e4800-d7ff-11ea-8f24-95426e5fbe9d.png)

I'd ideally like to keep the initialisation of the app gateway in Pulumi, since I want to control the SSL certs, front-end IP and scaling from there. Have added the code for my "blank" app gateway below, for reference

```
            var publicIpName = $"{context.StackName}-primary-aks-ip";
            var publicIpArgs = new PublicIpArgs
            {
                ResourceGroupName = builderWithNetwork.GetMainResourceGroupName(),
                Name = publicIpName,
                Sku = "Standard",
                AllocationMethod = "Static",
                IdleTimeoutInMinutes = 4,
                IpVersion = "IPv4",
            };
            var ipAddress = new PublicIp(publicIpName, publicIpArgs, customResourceOptions);

            var httpProbe = _GetDefaultHttpProbe();
            var httpsProbe = _GetDefaultHttpsProbe();
            var httpPort = new ApplicationGatewayFrontendPortArgs {Name = "httpPort", Port = 80};
            var httpsPort = new ApplicationGatewayFrontendPortArgs {Name = "httpsPort", Port = 443};
            var frontendIpConfiguration = new ApplicationGatewayFrontendIpConfigurationArgs(){ PublicIpAddressId = ipAddress.Id, Name = "appGatewayFrontendIP" };
            var defaultListener = new ApplicationGatewayHttpListenerArgs()
            {
                Name = "default-http-listener",
                Protocol = "Http",
                FrontendPortName = httpPort.Name,
                FrontendIpConfigurationName = frontendIpConfiguration.Name,
            };

            var defaultHttpSetting = new ApplicationGatewayBackendHttpSettingArgs()
            {
                Name = "defaulthttpsetting",
                Port = 80,
                CookieBasedAffinity = "Disabled",
                PickHostNameFromBackendAddress = false,
                Protocol = "Http",
                RequestTimeout = 30,
                ProbeName = httpProbe.Name
            };

            var defaultBackEndPool = new ApplicationGatewayBackendAddressPoolArgs()
            {
                Name = "defaultaddresspool",
                IpAddresses = "1.1.1.1" //Something that doesn't exist
            };

            var applicationGatewayRequestRoutingRuleArgs = new ApplicationGatewayRequestRoutingRuleArgs()
            {
                Name = "default-rule",
                HttpListenerName = defaultListener.Name,
                BackendAddressPoolName = defaultBackEndPool.Name,
                BackendHttpSettingsName = defaultHttpSetting.Name,
                RuleType = "Basic"
            };

            var frontEndCertificate = new ApplicationGatewaySslCertificateArgs
            {
                Name = "accurx-dot-com-wildcard-certificate",
                KeyVaultSecretId = "SECRET_ID_HERE",
            };

            var aksAppGateway = new ApplicationGatewayArgs
            {
                Name = $"{context.StackName}-primary-aks-appgatway",
                ResourceGroupName = builderWithNetwork.GetMainResourceGroupName(),
                Sku = new ApplicationGatewaySkuArgs
                {
                    Name = "WAF_v2",
                    Tier = "WAF_v2",
                },
                EnableHttp2 = false,
                Identity = new ApplicationGatewayIdentityArgs
                {
                    Type = "UserAssigned",
                    IdentityIds = builderWithNetwork.Names.BuildAppGatewayIdentity()
                },
                AutoscaleConfiguration = new ApplicationGatewayAutoscaleConfigurationArgs
                {
                    MaxCapacity = 10,
                    MinCapacity = 1
                },
                WafConfiguration = new ApplicationGatewayWafConfigurationArgs
                {
                    Enabled = true,
                    FirewallMode = "Detection",
                    Exclusions = new InputList<ApplicationGatewayWafConfigurationExclusionArgs>(),
                    RuleSetVersion = "3.0",
                    RuleSetType = "OWASP",
                    RequestBodyCheck = true,
                    FileUploadLimitMb = 100,
                    MaxRequestBodySizeKb = 128,
                },
                SslPolicies = new ApplicationGatewaySslPolicyArgs { PolicyType = "Predefined", PolicyName = "AppGwSslPolicy20170401S" },
                FrontendIpConfigurations = frontendIpConfiguration,
                GatewayIpConfigurations = new ApplicationGatewayGatewayIpConfigurationArgs {Name = "appGatewayIpConfig", SubnetId = builderWithNetwork.Network.DmzSubnet.Id},
                FrontendPorts = new [] { httpPort, httpsPort },
                Probes = new [] { httpProbe, httpsProbe },
                SslCertificates = new []{frontEndCertificate},

                //Below are set as defaults, then managed by AGIC
                BackendHttpSettings = defaultHttpSetting,
                HttpListeners = defaultListener,
                RequestRoutingRules = applicationGatewayRequestRoutingRuleArgs,
                BackendAddressPools = defaultBackEndPool,
            };

            var appGateway = new ApplicationGateway($"{context.StackName}-primary-aks-appgatway", aksAppGateway, new CustomResourceOptions()
            {
                Parent = this,
                IgnoreChanges = new List<string>()
                {
                    "backendAddressPools",
                    "requestRoutingRules",
                    "backendHttpSettings",
                    "httpListeners",
                }
            });
        }

        private static ApplicationGatewayProbeArgs _GetDefaultHttpsProbe()
        {
            return new ApplicationGatewayProbeArgs
            {
                Interval = 30,
                Timeout = 30,
                MinimumServers = 0,
                Name = "defaultprobe-Https",
                Protocol = "Https",
                Host = "localhost",
                UnhealthyThreshold = 3,
                Path = "/",
                PickHostNameFromBackendHttpSettings = false,
            };
        }

        private static ApplicationGatewayProbeArgs _GetDefaultHttpProbe()
        {
            return new ApplicationGatewayProbeArgs
            {
                Interval = 30,
                Timeout = 30,
                MinimumServers = 0,
                Name = "defaultprobe-Http",
                Protocol = "Http",
                Host = "localhost",
                UnhealthyThreshold = 3,
                Path = "/",
                PickHostNameFromBackendHttpSettings = false,
            };
        }
```

**Comment by lukehoban** [comment]
Just to make sure I understand correctly - things work fine as long as you don't do a refresh, and doing the refresh itself succeeds, but doing a preview (or presumably update) after the refresh fails?

cc also @mikhailshilkov

**Comment by lbargery** [comment]
Yep that’s exactly right...I could get away without doing the refresh but obviously not ideal!

**Comment by ringods** [comment]
I guess this would require AGIC support in the upstream TF provider to become available.

https://github.com/terraform-providers/terraform-provider-azurerm/issues/7384

**Comment by lbargery** [comment]
Is there any way I can use "ignore" to make it work for now? When I ignored the HttpListeners etc it didn't seem to work

**Comment by Aaronontheweb** [comment]
@lbargery I ran into some issues using Ignore on AGIC last night myself: https://stackoverflow.com/questions/63838596/pulumi-how-to-ignore-changes-made-to-previously-deployed-resources

Same types of errors you're reporting here

**Comment by lbargery** [comment]
Can we get an update on whether we can use ignore to at least make this work? It's quite problematic not being able to do a stack refresh at any point now (without lots of manual editing)

**Comment by lbargery** [comment]
@mikhailshilkov is there any way that you could add a new parameter to "Pulumi refresh" to exlude certain URNs. I notice that you have ```-t, --target stringArray``` as an option? A similar one for exclude could be handy here

**Comment by Aaronontheweb** [comment]
Any updates on this? Just created an outage on the service we're using with Pulumi because of this change - accidentally wiped away all of our backend targets we use for traffic routing when we were changing a different part of the same stack.

**Comment by felipebrito1** [comment]
This also happened to me after publishing a net core 5.0 application in an AppService resource via Pipeline AzureDevops.
Changes to the following resources have occurred after "pulumi preview" command:

**AppServicePlan**: 
~ numberOfSites            : 0 => 1

**AppService**: 
~ appSettings                   : { 
 + ConnectionStrings__xxxx: "xxx" 
}
~ siteConfig                    : {
  + linuxFxVersion         : "DOTNETCORE|5.0"
  + remoteDebuggingVersion : "VS2019"
  ~ scmType                : "None" => "VSTSRM"
}
- siteCredentials               : [
      -     [0]: {
              - password: "[secret]"
              - username: "xxx"
            }
        ]
+ siteCredentials               : "[secret]"
~ sourceControl                 : {
	branch           : "master"
	manualIntegration: false
  + repoUrl          : "VSTSRM"
	rollbackEnabled  : false
	useMercurial     : false
}

**Comment by mikhailshilkov** [comment]
The upstream issue referred [above](https://github.com/pulumi/pulumi-azure/issues/655#issuecomment-688096526) has been fixed a long time ago and shipped in 4.1.0 of this provider. There haven't been any comments since then, so I'll assume it works now and the issue can be closed as resolved. Please report back if it's not.
